### ISSUE TITLE
strncmp declaration error on Unbuntu 14.04 (Trusty)

### ISSUE THREAD

**Issue by lenary** [Microsoft/checkedc#133]
Our declaration for `strncmp(s1, s2, n)` on Ubuntu 14.04 is causing errors.

This is because Ubuntu overrides the declaration of strncmp with the following macro:
```
/* Compare N characters of S1 and S2.  */
#ifndef _HAVE_STRING_ARCH_strncmp
# define strncmp(s1, s2, n)                                                   \
  (__extension__ (__builtin_constant_p (n)                                    \
                  && ((__builtin_constant_p (s1)                              \
                       && strlen (s1) < ((size_t) (n)))                       \
                      || (__builtin_constant_p (s2)                           \
                          && strlen (s2) < ((size_t) (n))))                   \
                  ? strcmp (s1, s2) : strncmp (s1, s2, n)))
#endif
```

This  seems a lot harder to solve than what we did for apple's system headers.

**Comment by lenary** [comment]
I think the solution might be heavy use of #ifdef. Something like

```
#ifdef strncmp
#define _strncmp_macro(s1, s2, n) strncmp(s1, s2, n)
#undef strncmp
#endif
int strncmp(...);
#ifdef _strncmp_macro
#define strncmp(s1, s2, n) _strncmp_macro(s1, s2, n)
#undef _strncmp_macro
#endif
```

I have no idea if this will work, but worth a try, right? I think this is slightly preferable to just not defining the safe interface.

**Comment by dtarditi** [comment]
This was fixed by PR #168.  The solution was to `#undef` the macro, following suggestions from the C standard on how to handle this sort of situation.    This might disable optimizations in clang that relied on the macro, but it is correct.
